dbms/src/Storages/DeltaMerge/tests/gtest_dm_segment.cpp

```diff
@@ -1024,6 +1026,142 @@ try
+        auto read_ranges = {RowKeyRange::newAll(false, 1)};
+        auto real_ranges = segment->shrinkRowKeyRanges(read_ranges);
+        auto bitmap_filter = segment->buildBitmapFilter( //
+            dmContext(),
+            segment_snap,
+            real_ranges,
+            EMPTY_RS_OPERATOR,
+            std::numeric_limits<UInt64>::max(),
+            DEFAULT_BLOCK_SIZE);
+        // the bitmap only contains the overlapped packs of ColumnFileBig. So only 60 here.
+        ASSERT_EQ(bitmap_filter->size(), 60);
+        ASSERT_EQ(bitmap_filter->toDebugString(), "000000000011111111111111111111111111111111111111110000000000");
+    }
+
+    ColumnDefines cols_to_read{
+        getExtraHandleColumnDefine(false),
+    };
+    {
+        // test read data
+        auto segment_snap = segment->createSnapshot(dmContext(), false, CurrentMetrics::DT_SnapshotOfRead);
+        auto in = segment->getBitmapFilterInputStream(
+            dmContext(),
+            cols_to_read,
+            segment_snap,
+            {RowKeyRange::newAll(false, 1)},
+            EMPTY_FILTER,
+            std::numeric_limits<UInt64>::max(),
+            DEFAULT_BLOCK_SIZE,
+            DEFAULT_BLOCK_SIZE);
+        ASSERT_INPUTSTREAM_BLOCK_UR(
+            in,
+            Block({
+                createColumn<Int64>(createNumbers<Int64>(30, 70)),
+            }));
+    }
+
+    {
+        // let segment's rowkey_range be [30, 90)
+        HandleRange range(30, 90);
+        segment->rowkey_range = RowKeyRange::fromHandleRange(range);
+        // delete range [50, 80)
+        Block block = DMTestEnv::prepareSimpleWriteBlock(
+            50,
+            80,
+            false,
+            3,
+            "_tidb_rowid",
+            EXTRA_HANDLE_COLUMN_ID,
+            EXTRA_HANDLE_COLUMN_INT_TYPE,
+            false,
+            1,
+            true,
+            /*is_deleted=*/true);
+        segment->write(dmContext(), std::move(block));
+        // write range [80, 90)
+        Block block2 = DMTestEnv::prepareSimpleWriteBlock(80, 90, false);
+        segment->write(dmContext(), std::move(block2));
+    }
+    {
+        // test read data with delete-range and new writes
+        auto segment_snap = segment->createSnapshot(dmContext(), false, CurrentMetrics::DT_SnapshotOfRead);
+        auto in = segment->getBitmapFilterInputStream(
+            dmContext(),
+            cols_to_read,
+            segment_snap,
+            {RowKeyRange::newAll(false, 1)},
+            EMPTY_FILTER,
+            std::numeric_limits<UInt64>::max(),
+            DEFAULT_BLOCK_SIZE,
+            DEFAULT_BLOCK_SIZE);
+        // Only the rows in [30, 50) and [80, 90) valid
+        auto vec = createNumbers<Int64>(30, 50);
+        vec.append_range(createNumbers<Int64>(80, 90));
+        ASSERT_INPUTSTREAM_BLOCK_UR(
+            in,
+            Block({
+                createColumn<Int64>(vec),
+            }));
```

Comment:
I've pushed a commit to ensure the read rows values are correct. /cc @Lloyd-Pottiger